Parallelize tests only when there are enough to justify the parallelization overhead

jorgemanrubia opened the pull request:
Parallelizing tests has a cost in terms of database setup and fixture loading. This change  makes Rails disable parallelization when the number of tests is below a configurable threshold.

When running tests in parallel each process gets its own database instance. On each execution, each process will update each database schema (if needed) and load all the fixtures. This can be very expensive for non trivial datasets. 

As an example, for HEY, when running a single file with 18 tests, running tests in parallel in my box adds an overhead of 13 seconds versus not parallelizing them. Of course parallelizing is totally worthy when there are many tests to run, but not when running just a few tests.

The threshold is configurable via `config.active_support.test_parallelization_minimum_number_of_tests`, which is ~30~ 50 by default. 

This also adds some tracing to know how tests are being executed:

When in parallel:

```
Running 2829 tests in parallel using 8 processes
```

When not in parallel:

```
Running 15 tests in a single process (parallelization threshold is 30)
```

Pending:

- [x] Document
- [x] Changelog
- [x] Tests
- [x] Refine default (30 is pretty much a wild guess). Update: [see this](https://github.com/rails/rails/pull/42761#issuecomment-879795839).

cc @eileencodes

jorgemanrubia committed:
Parallelize tests only when overhead is justified

Parallelizing tests has a cost in terms of database setup and fixture
loading. This change makes Rails disable parallelization when the number
of tests is below a configurable threshold.

When running tests in parallel each process gets its own database
instance. On each execution, each process will update each database
schema (if needed) and load all the fixtures. This can be very expensive
for non trivial datasets.

As an example, for HEY, when running a single file with 18 tests,
running tests in parallel in my box adds an overhead of 13 seconds
versus not parallelizing them. Of course parallelizing is totally worthy
when there are many tests to run, but not when running just a few tests.

The threshold is configurable via
config.active_support.test_parallelization_minimum_number_of_tests,
which is 30 50 by default.

This also adds some tracing to know how tests are being executed:

When in parallel:

```
Running 2829 tests in parallel in 8 processes
```

When not in parallel:

```
Running 15 tests in a single process (parallelization threshold is 30)
```

jorgemanrubia commented on activesupport/lib/active_support/testing/parallelize_when_suitable_executor.rb:
```diff
@@ -0,0 +1,37 @@
+      end
+
+      def start
+        parallelize if worth_parallelizing?
```
I moved the parallelization logic here because we need to run it **after** all the test files are loaded. I did it this way to avoid having to change/patch Minitest. This is just a way of loading the parallelization logic in a lazy way, after all the test cases have been loaded. Happy to change the approach if you can think of a better one.

eileencodes replied:
I think I prefer `should_parallelize` over `worth_parallelizing`. "worthy" feels like a value judgement to me.

ghiculescu commented:
This would need a changelog entry. I'm a big fan of the concept.

jorgemanrubia commented:
Yes @ghiculescu it's there in the "pending" list. I'll add :)

eileencodes commented on activesupport/lib/active_support/test_case.rb:
```diff
@@ -88,9 +89,7 @@ def parallelize(workers: :number_of_processors, with: :processes)
-        Minitest.parallel_executor = executor
-
-        parallelize_me!
+        Minitest.parallel_executor = Testing::ParallelizeWhenSuitableExecutor.new(executor)
```
Do we have access to runnables before this? I think it makes more sense to not even create the executor if we don't have enough tests. Then the check could be moved up around line 79 where we return if workers are less than or equal to 1. I don't love the `WhenSuitableExecutor` name either and we could avoid renaming it by moving the check.

jorgemanrubia replied:
@eileencodes yes that was my initial idea but, unfortunately, when `parallelize` runs, test cases aren't loaded yet, so you can't really count tests and make decisions. The executor path is not ideal, but we need some kind of lazy mechanism invoked from `Minitest.run`. I didn't want to extend/modify Minitest for this, so ended up with this approach as the best one I could find.

Happy to rename `ParallelizeWhenSuitableExecutor`. `ConditionalParallelExecutor` or something like that?

jorgemanrubia replied:
I added some code to show a trace line when running tests with info on how they ran:

When in parallel:

```
Running 2829 tests in parallel in 8 processes
```

When not in parallel:

```
Running 15 tests in a single process (parallelization threshold is 20)
```

eileencodes commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+*   Avoid parallelizing tests when only running a few
+
+    Running tests in parallel adds an important overhead in terms of database
```
```suggestion
    Running tests in parallel adds overhead in terms of database
```

eileencodes reviewed (commented):
I don't know how I feel about this feature as it silently changes behavior for existing applications. I think if someone isn't explicitly turning off parallelization then there should be a warning that parallelization isn't turned on.

eileencodes reviewed (commented):
This is looking good, I like that new warning and the method name changes. I think `ParallelizeWhenSuitableExecutor` should just be `ParallelizeExecutor` since it always gets initialized. Once that's fixed, can you also squash your commits into one?

eileencodes commented on activesupport/lib/active_support/testing/parallelize_when_suitable_executor.rb:
```diff
@@ -0,0 +1,53 @@
+
+        def execution_info
+          if should_parallelize?
+            executor_label = @parallel_executor.is_a?(ActiveSupport::Testing::Parallelization) ? "processes" : "threads"
```
Instead of doing an `is_a` here what do you think about passing the option along with `initialize`?

eileencodes commented on activesupport/lib/active_support/testing/parallelize_when_suitable_executor.rb:
```diff
@@ -0,0 +1,53 @@
+
+module ActiveSupport
+  module Testing
+    class ParallelizeWhenSuitableExecutor
```
This should be no-doc'd

```suggestion
    class ParallelizeExecutor # :nodoc:
```

eileencodes reviewed (commented):
I have a couple more comments after I hit enter on that last one, sorry for the notification noise.

I think that parallelization should always be on if `PARALLEL_WORKERS` is set above 1 because that means it's an explicit setting and I think it would be odd to have to change a config and an env var.

jorgemanrubia commented:
@eileencodes I think I addressed your comments. I changed things a bit to move some configuration decisions into the new `ParalellizeExecutor`.

I still have pending to add some tests for this and wrap things. I'll do that today or tomorrow. 

Thanks for the review and feedback 🙏.

jorgemanrubia commented:
@dhh had suggested the possibility of adapting the number of workers to the number of tests. E.g: 30 tests, 1 worker. 60 tests, 2 workers, etc. I think it would be great and it would be totally feasible with this implementation. I'd leave that for a future iteration though.

jorgemanrubia commented:
I wrote [this script](https://gist.github.com/jorgemanrubia/32e031fa2d5911a1d839739aeb43ecee) for trying to determine a right default threshold here. It measures and compares times for running tests in parallel versus running them sequentially. When running the script with HEY and Basecamp I got 79 and 85, respectively.

For measuring, I established a synthetic test duration of 100ms each test. If no delay, threshold goes ridiculously high (it's always faster to run sequentially).

I lean towards setting the default threshold conservatively at 50.

jorgemanrubia commented:
This is good to test from my side @eileencodes
